File: tools/HME/scripts/hme_paths.py
from __future__ import annotations

import os
import re
from pathlib import Path

PROJECT_ROOT = Path(os.environ.get("PROJECT_ROOT") or Path(__file__).resolve().parents[3]).resolve()
_ENV_REF_RE = re.compile(r"\$\{([A-Za-z_][A-Za-z0-9_]*)\}|\$([A-Za-z_][A-Za-z0-9_]*)")


def _expand_env_refs(value: str) -> str:
    out = value
    for _ in range(4):
        nxt = _ENV_REF_RE.sub(lambda m: os.environ.get(m.group(1) or m.group(2), m.group(0)), out)
        if nxt == out:
            break
        out = nxt
    return out


def _under_root(value: str | None, fallback: Path) -> Path:
    if not value:
        return fallback
    expanded = _expand_env_refs(str(value))
    if _ENV_REF_RE.search(expanded):
        return fallback
    candidate = Path(expanded).expanduser().resolve()
    try:
        candidate.relative_to(PROJECT_ROOT)
        return candidate
    except ValueError:
        return fallback


HME_RUNTIME_DIR = _under_root(
    os.environ.get("HME_RUNTIME_DIR"), PROJECT_ROOT / "tools" / "HME" / "runtime"
)
HME_METRICS_DIR = _under_root(os.environ.get("HME_METRICS_DIR"), HME_RUNTIME_DIR / "metrics")
HME_STATE_DIR = _under_root(os.environ.get("HME_STATE_DIR"), HME_RUNTIME_DIR / "state")
COMPOSITION_OUTPUT_DIR = _under_root(
    os.environ.get("COMPOSITION_OUTPUT_DIR"), PROJECT_ROOT / "src" / "output"
)
COMPOSITION_METRICS_DIR = _under_root(
    os.environ.get("COMPOSITION_METRICS_DIR") or os.environ.get("METRICS_DIR"),
    COMPOSITION_OUTPUT_DIR / "metrics",
)

PROJECT_METRIC_NAMES = {
    "adaptive-state.json",
    "composition-diff.json",
    "composition-diff.md",
    "current-run.json",
    "feedback_graph.json",
    "fingerprint-comparison.json",
    "golden-fingerprint.json",
    "golden-fingerprint.prev.json",
    "hci-snapshot-diff.json",
    "hci-verifier-snapshot.json",
    "journal.md",
    "l0-dump.json",
    "narrative-digest.md",
    "perceptual-report.json",
    "pipeline-summary.json",
    "run-comparison.json",
    "runtime-snapshots.json",
    "system-manifest.json",
    "trace-summary.json",
    "trace.jsonl",
    "verdict-model.json",
}

HME_METRIC_NAMES = {
    "detector-stats.jsonl",
    "hci-regression-alert.json",
    "kb-signatures.json",
    "kb-staleness.json",
    "kb-trust-weights.json",
    "legacy-override-history.jsonl",
    "mode-classifier.jsonl",
    "reflections.jsonl",
    "satisfaction.jsonl",
    "todo-graph.md",
    "vram-history.jsonl",
}


def _name(parts: tuple[str, ...]) -> str:
    return str(parts[0]) if parts else ""


def is_hme_metric_name(*parts: str) -> bool:
    name = _name(parts)
    if not name or name in PROJECT_METRIC_NAMES or name == "run-history":
        return False
    return name.startswith("hme-") or name in HME_METRIC_NAMES


def hme_metric(*parts: str) -> Path:
    return HME_METRICS_DIR.joinpath(*parts)


def hme_state(*parts: str) -> Path:
    return HME_STATE_DIR.joinpath(*parts)


def project_metric(*parts: str) -> Path:
    return COMPOSITION_METRICS_DIR.joinpath(*parts)


def metric_path(*parts: str) -> Path:
    return hme_metric(*parts) if is_hme_metric_name(*parts) else project_metric(*parts)


def read_hme_metric(*parts: str) -> Path:
    primary = hme_metric(*parts)
    return primary if primary.exists() else project_metric(*parts)


def read_metric_path(*parts: str) -> Path:
    return read_hme_metric(*parts) if is_hme_metric_name(*parts) else project_metric(*parts)


def write_hme_metric(*parts: str) -> Path:
    return hme_metric(*parts)


def write_project_metric(*parts: str) -> Path:
    return project_metric(*parts)


def write_metric_path(*parts: str) -> Path:
    return metric_path(*parts)

File: tools/HME/proxy/hme_paths.js
'use strict';
const { requireEnv: _hmeRequireEnv } = require('./shared/load_env.js');

const fs = require('fs');
const path = require('path');

const PROJECT_ROOT = _hmeRequireEnv('PROJECT_ROOT');

const ENV_REF_RE = /\$\{([A-Za-z_][A-Za-z0-9_]*)\}|\$([A-Za-z_][A-Za-z0-9_]*)/g;

function hasEnvRef(value) {
  return ENV_REF_RE.test(String(value || ''));
}

function expandEnvRefs(value) {
  let out = String(value || '');
  for (let i = 0; i < 4; i += 1) {
    ENV_REF_RE.lastIndex = 0;
    const next = out.replace(ENV_REF_RE, (match, braced, bare) => {
      const replacement = process.env[braced || bare];
      return replacement ? String(replacement) : match;
    });
    if (next === out) break;
    out = next;
  }
  return out;
}

function underRoot(candidate, fallback) {
  const absRoot = path.resolve(PROJECT_ROOT);
  if (!candidate) return fallback;
  const expanded = expandEnvRefs(candidate);
  if (hasEnvRef(expanded)) return fallback;
  const abs = path.resolve(expanded);
  if (abs === absRoot || abs.startsWith(absRoot + path.sep)) return abs;
  return fallback;
}

const HME_RUNTIME_DIR = underRoot(
  process.env.HME_RUNTIME_DIR,
  path.join(PROJECT_ROOT, 'tools', 'HME', 'runtime'),
);
const HME_METRICS_DIR = underRoot(
  process.env.HME_METRICS_DIR,
  path.join(HME_RUNTIME_DIR, 'metrics'),
);
const HME_STATE_DIR = underRoot(
  process.env.HME_STATE_DIR,
  path.join(HME_RUNTIME_DIR, 'state'),
);
const COMPOSITION_OUTPUT_DIR = underRoot(
  process.env.COMPOSITION_OUTPUT_DIR,
  path.join(PROJECT_ROOT, 'src', 'output'),
);
const COMPOSITION_METRICS_DIR = underRoot(
  _hmeRequireEnv('COMPOSITION_METRICS_DIR'),
  path.join(COMPOSITION_OUTPUT_DIR, 'metrics'),
);

const PROJECT_METRIC_NAMES = new Set([
  'adaptive-state.json',
  'composition-diff.json',
  'composition-diff.md',
  'current-run.json',
  'feedback_graph.json',
  'fingerprint-comparison.json',
  'golden-fingerprint.json',
  'golden-fingerprint.prev.json',
  'hci-snapshot-diff.json',
  'hci-verifier-snapshot.json',
  'journal.md',
  'l0-dump.json',
  'narrative-digest.md',
  'perceptual-report.json',
  'pipeline-summary.json',
  'run-comparison.json',
  'runtime-snapshots.json',
  'system-manifest.json',
  'trace-summary.json',
  'trace.jsonl',
  'verdict-model.json',
]);

const HME_METRIC_NAMES = new Set([
  'detector-stats.jsonl',
  'hci-regression-alert.json',
  'kb-signatures.json',
  'kb-staleness.json',
  'kb-trust-weights.json',
  'legacy-override-history.jsonl',
  'mode-classifier.jsonl',
  'reflections.jsonl',
  'satisfaction.jsonl',
  'todo-graph.md',
  'vram-history.jsonl',
]);

function metricName(parts) { return parts.length > 0 ? String(parts[0]) : ''; }

function isHmeMetricName(...parts) {
  const name = metricName(parts);
  if (!name || PROJECT_METRIC_NAMES.has(name) || name === 'run-history') return false;
  return name.startsWith('hme-') || HME_METRIC_NAMES.has(name);
}

function hmeMetric(...parts) { return path.join(HME_METRICS_DIR, ...parts); }
function hmeState(...parts) { return path.join(HME_STATE_DIR, ...parts); }
function projectMetric(...parts) { return path.join(COMPOSITION_METRICS_DIR, ...parts); }
function metricPath(...parts) { return isHmeMetricName(...parts) ? hmeMetric(...parts) : projectMetric(...parts); }

function readHmeMetric(...parts) {
  const primary = hmeMetric(...parts);
  return fs.existsSync(primary) ? primary : projectMetric(...parts);
}

function readMetricPath(...parts) {
  return isHmeMetricName(...parts) ? readHmeMetric(...parts) : projectMetric(...parts);
}

function writeHmeMetric(...parts) { return hmeMetric(...parts); }
function writeProjectMetric(...parts) { return projectMetric(...parts); }
function writeMetricPath(...parts) { return metricPath(...parts); }

module.exports = {
  PROJECT_ROOT,
  HME_RUNTIME_DIR,
  HME_METRICS_DIR,
  HME_STATE_DIR,
  COMPOSITION_OUTPUT_DIR,
  COMPOSITION_METRICS_DIR,
  hmeMetric,
  hmeState,
  projectMetric,
  metricPath,
  readHmeMetric,
  readMetricPath,
  writeHmeMetric,
  writeProjectMetric,
  writeMetricPath,
  isHmeMetricName,
};
